cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,948 @@
+    : data_{[&res, &other, specified_mem_type, this]() {
+        auto other_mem_type = other.mem_type();
+        auto mem_type       = specified_mem_type.value_or(other_mem_type);
+        auto result         = storage_type_variant{};
+        if (mem_type == other.mem_type()) {
+          result = std::move(other.data_);
+        } else if (!other.is_owning() && has_compatible_accessibility(other_mem_type, mem_type) &&
+                   !is_host_device_accessible(mem_type)) {
+          switch (mem_type) {
+            case (memory_type::host): {
+              result = std::visit(
+                [&result, this](auto&& other_view) {
+                  return view_type<memory_type::host>{
+                    other_view.data_handle(),
+                    other_view.mapping(),
+                    cp_.template make_accessor_policy<memory_type::host>()};
+                },
+                other.view());
+              break;
+            }
+            case (memory_type::device): {
+              result = std::visit(
+                [&result, this](auto&& other_view) {
+                  return view_type<memory_type::device>{
+                    other_view.data_handle(),
+                    other_view.mapping(),
+                    cp_.template make_accessor_policy<memory_type::device>()};
+                },
+                other.view());
+              break;
+            }
+            case (memory_type::managed): {
+              result = std::visit(
+                [&result, this](auto&& other_view) {
+                  return view_type<memory_type::managed>{
+                    other_view.data_handle(),
+                    other_view.mapping(),
+                    cp_.template make_accessor_policy<memory_type::managed>()};
+                },
+                other.view());
+              break;
+            }
+            case (memory_type::pinned): {
+              result = std::visit(
+                [&result, this](auto&& other_view) {
+                  return view_type<memory_type::pinned>{
+                    other_view.data_handle(),
+                    other_view.mapping(),
+                    cp_.template make_accessor_policy<memory_type::pinned>()};
+                },
+                other.view());
+              break;
+            }
+          }
+        } else {
+          result = copy_from(res, other, mem_type);
+        }
+        return result;
+      }()}
+  {
+  }
+
+  /**
+   * @brief Construct one mdbuffer from another mdbuffer lvalue with matching
+   * element type, extents, layout, and container policy.
+   *
+   * Unlike when constructing from an rvalue, the new mdbuffer will take a
+   * non-owning view whenever possible, since it is assumed that the caller
+   * will manage the lifetime of the lvalue input.
+   */
+  mdbuffer(raft::resources const& res,
+           mdbuffer<ElementType, Extents, LayoutPolicy, ContainerPolicy>& other,
```

Comment:
Should the `other` be here a `const mdbuffer&`?

Reply:
No, this constructor should only apply to non-const lvalues. The `mdbuffer` constructors are all intended to take some data and make it accessible in exactly the same format modulo whatever changes you explicitly asked for. If `other` were passed as a const lvalue reference, we could only obtain const views from it (i.e. mdspans with const element type), so the mdbuffer constructed from it would have const element type rather than the original type passed in.

In this case, the caller almost certainly means to have constructed an mdbuffer of const _data_ rather than a const mdbuffer. If we provided an overload for const lvalue reference, it would also be ambiguous whether the caller was tryng to construct that non-owning mdbuffer of const data or if they actually wanted to perform a copy of the (possibly) underlying mdarray.

By simply not supporting a constructor from a const lvalue reference, we avoid a bunch of potential confusion and encourage users to use mdbuffer as a highly temporary object, relying on mdspans obtained from the mdbuffer for actual data manipulation.

Callers can still work with const lvalue references to mdbuffers if necessary, but they will need to be explicit about how the ambiguity should be resolved. For instance, if they want to allow a copy of the (possibly) underlying mdarray, they can make use of the default copy constructor:
```
void foo(raft::resources const& res, mdbuffer const& buf) {
  // Explicit use of copy constructor, which behaves according to familiar rules for mdspan and mdarray copies
  auto dev_buf = mdbuffer(res, mdbuffer(buf), memory_type::device);
}

void bar(raft::resources const& res, mdbuffer const& buf) {
  auto dev_buf = std::visit(
    [](auto view) {
      // mdbuffer constructed from a view of const data following familiar rules of mdspan-to-mdbuffer construction
      return mdbuffer(res, view, memory_type::device);
    },
    buf.view()
  );
}
```

Regardless of which the user chooses, it forces them to be explicit about the behavior they expect if they really, really, really meant to use a const mdbuffer and not an mdbuffer of const data.

Reply:
Ok, container const-ness is always a little bit confusing, but I think I've got your point :)
Maybe just add a short `/* NOLINT */` to the argument, so that the reader sees this is intentional?

Reply:
Just did so and updated the docs in a few places. I agree on container const-ness being tricky! If you hadn't asked the question above about converting to const views without going through `make_const_mdspan`, I wouldn't have realized that the technique I expected people to use for _this_ question (const element type) wasn't working. Thanks very much for pushing me on both these!